### PR TITLE
chore: replace external cypress/parse-domain repo

### PR DESCRIPTION
**Pull request by jennifer-shehane**
### Additional details

- Replaces the **`@cypress/parse-domain`** dependency in **`@packages/network-tools`** with **`tldts`**, wrapped in **`lib/parse-domain.ts`** so **`parseDomain`** still returns the legacy `{ subdomain, domain, tld } | null` shape used by CORS, `document.domain` injection, and server cookie logic.
- `@cypress/parse-domain` is an old fork (last published 2020) and pulled in an outdated dependency chain (e.g. **`got@8`**). **`tldts`** is actively maintained, ships TypeScript and a single CJS entry suitable for isomorphic use, and aligns with the public suffix list plus private rules needed for Cypress’s origin / super-domain / cookie-domain behavior.

## Risks / follow-ups
- **PSL drift**: Upgrading **`tldts`** can slightly change edge-case behavior vs the frozen 2020 trie; existing tests should catch regressions for supported cases.
- **`customTlds` as `RegExp`**: If any external caller relied on regex-based custom suffixes, consider a follow-up (e.g. map **`string[]`** to **`validHosts`**).

---

> [!NOTE]
> **Medium Risk**
> Changes public-suffix and hostname parsing used for CORS, super-domain origins, and cookie-domain behavior; behavior is intentionally preserved but PSL/tldts upgrades could shift edge cases.
> 
> **Overview**
> **`@packages/network-tools`** drops **`@cypress/parse-domain`** (and its **`got@8`** dependency chain) in favor of **`tldts`**, with a new **`lib/parse-domain.ts`** that still exposes **`parseDomain`** as **`{ subdomain, domain, tld } | null`**.
> 
> **`cors.ts`** now imports that local helper instead of wrapping the old package; **`parse-domain`** is re-exported from the package entry. The wrapper maps **`tldts`** results into the legacy shape, keeps **`privateTlds: true`** by default, and preserves prior edge cases (leading-dot cookie domains, bare **`localhost`** → null for caller fallback, IPv4/IPv6, legacy **`customTlds`**-style splits like **`12.`**).
> 
> New vitest coverage documents those compatibility behaviors. **`yarn.lock`** shrinks from removing transitive deps of **`@cypress/parse-domain`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 710129d25e6f93af96f03976008d319fa89d8425. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### Steps to test

- [ ] `yarn workspace @packages/network-tools test`
- [ ] `yarn workspace @packages/server test-unit -- cdp_automation_spec.ts`
- [ ] `yarn workspace @packages/proxy test -- test/unit/http/request-middleware.spec.ts`
- [ ] `yarn workspace @packages/proxy test -- test/unit/http/response-middleware.spec.ts`

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/network-tools/lib/parse-domain.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+export type ParsedDomainParts = {
+  subdomain: string
+  domain: string
+  tld: string
+}
+
+export type ParseDomainOptions = {
+  privateTlds?: boolean
+  /** Ignored for compatibility with the old parse-domain API; unknown suffixes use tldts plus the legacy default `customTlds` rule from the former cors wrapper. */
+  customTlds?: RegExp | string[]
```
Comment (AtofStryker) [inline packages/network-tools/lib/parse-domain.ts]: Exporting `customTlds` in the type signature but silently ignoring it is a footgun. The only known caller of `customTlds` was the internal `cors.ts` wrapper, which is now gone. Consider removing the field from `ParseDomainOptions` entirely — it eliminates the option-that-does-nothing, the compatibility JSDoc, and the test below that verifies a no-op.

File: packages/network-tools/lib/parse-domain.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ * has no PSL row (empty `publicSuffix`).
+ */
+const LEGACY_CUSTOM_TLDS_RE = /(^[\d.]+$|\.[^.]+$)/
+
+function legacyCustomTldSplit (hostname: string): ParsedDomainParts | null {
+  if (!LEGACY_CUSTOM_TLDS_RE.test(hostname)) {
+    return null
+  }
+
+  if (/^[\d.]+$/.test(hostname)) {
```
Comment (AtofStryker) [inline packages/network-tools/lib/parse-domain.ts:39]: `tldtsParse` marks valid IPv4 addresses with `isIp: true`, so `tldtsToLegacy` handles them before this branch is ever reached. The only strings that reach `legacyCustomTldSplit` with a digit-dot match are malformed ones like `'12.'` — not real IPs. Either remove the IPv4 branch here and note the function only handles malformed digit-dot strings, or add a comment explaining when tldts would fail to detect `isIp` but still match `^[\d.]+$`.

File: packages/network-tools/lib/parse-domain.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+export type ParseDomainOptions = {
+  privateTlds?: boolean
+  /** Ignored for compatibility with the old parse-domain API; unknown suffixes use tldts plus the legacy default `customTlds` rule from the former cors wrapper. */
+  customTlds?: RegExp | string[]
+}
+
+const TLDT_OPTS_BASE = {
+  extractHostname: false,
```
Comment (AtofStryker) [inline packages/network-tools/lib/parse-domain.ts:18]: With `extractHostname: false`, passing a full URL (e.g. `'https://example.com'`) returns wrong results from tldts silently. All current callers pass hostnames, so this isn't a bug today — but since `parseDomain` is now a public export from the package, a short note here would help future callers avoid the trap.

File: packages/network-tools/test/unit/parse-domain.spec.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      expect(parseDomain('www.example.com', {
+        customTlds: /\.local$/,
+      })).toEqual({
+        subdomain: 'www',
+        domain: 'example',
+        tld: 'com',
+      })
+    })
+  })
+})
```
Comment (AtofStryker) [inline packages/network-tools/test/unit/parse-domain.spec.ts:127]: Missing test for FQDN trailing-dot hostnames (`'example.com.'`). The wrapper strips leading dots but not trailing ones. Cookie domain attributes and some DNS resolutions can produce trailing-dot form, and `tldts` behavior here may differ from `@cypress/parse-domain`. Worth a test to document expected behavior.

**Review by AtofStryker — COMMENTED** [review]
One additional Public Suffix List (PSL) note worth tracking: the old library's bundled PSL (frozen ~2020) and `tldts`'s current PSL differ on `railway.app`. The old library treated it as a private suffix, so `foo.railway.app` and `bar.railway.app` had different super-domains. `tldts` no longer lists it, so they now share super-domain railway.app. This change likely reflects Railway removing itself from the PSL rather than a `tldts` bug — but the practical blast radius is narrow: it only affects users testing against a live `*.railway.app` deployment (not localhost), with `injectDocumentDomain: true` enabled, and navigating between two different Railway subdomains in the same test. More of a "heads up if someone files a bug" note than a blocker.

File: packages/network-tools/lib/parse-domain.ts (hunk 1 above)
Comment (jennifer-shehane) [reply packages/network-tools/lib/parse-domain.ts]: Done: field, JSDoc, and no-op test removed. https://github.com/cypress-io/cypress/pull/33819/commits/6673acaa79174dd65ae4444cd99904022cb94388

File: packages/network-tools/lib/parse-domain.ts (hunk 2 above)
Comment (jennifer-shehane) [reply packages/network-tools/lib/parse-domain.ts:39]: Added a comment: real IPs go through tldts isIp; this only catches malformed digit-dot strings like '12.' https://github.com/cypress-io/cypress/pull/33819/commits/6673acaa79174dd65ae4444cd99904022cb94388

File: packages/network-tools/lib/parse-domain.ts (hunk 3 above)
Comment (jennifer-shehane) [reply packages/network-tools/lib/parse-domain.ts:18]: Added a note that callers must pass a hostname, not a full URL. https://github.com/cypress-io/cypress/pull/33819/commits/6673acaa79174dd65ae4444cd99904022cb94388

File: packages/network-tools/test/unit/parse-domain.spec.ts (hunk 4 above)
Comment (jennifer-shehane) [reply packages/network-tools/test/unit/parse-domain.spec.ts:127]: Added: parseDomain('example.com.') → null https://github.com/cypress-io/cypress/pull/33819/commits/6673acaa79174dd65ae4444cd99904022cb94388

File: packages/network-tools/lib/parse-domain.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ * defaults (private suffixes, IPs). Does not support arbitrary RegExp customTlds beyond
+ * what tldts + PSL already classifies.
+ */
+export function parseDomain (input: string, options: ParseDomainOptions = {}): ParsedDomainParts | null {
+  const merged: ParseDomainOptions & { privateTlds: boolean } = {
+    privateTlds: true,
+    ...options,
+  }
+
+  const hostname = input.replace(/^\./, '').trim()
```
[the remark this answers is not in the document]
Comment (jennifer-shehane) [reply packages/network-tools/lib/parse-domain.ts]: Fixed now trim().replace(/^\.+/, '') https://github.com/cypress-io/cypress/pull/33819/commits/6673acaa79174dd65ae4444cd99904022cb94388

File: packages/network-tools/lib/parse-domain.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    return null
+  }
+
+  const r = tldtsParse(hostname, {
+    ...TLDT_OPTS_BASE,
+    allowPrivateDomains: merged.privateTlds !== false,
+  })
+
+  return tldtsToLegacy(hostname, r)
+}
```
[the remark this answers is not in the document]
Comment (jennifer-shehane) [reply packages/network-tools/lib/parse-domain.ts:116]: Handled: the legacyCustomTldSplit fallback still matches non-PSL domains, so scoping is preserved. https://github.com/cypress-io/cypress/pull/33819/commits/6673acaa79174dd65ae4444cd99904022cb94388

**Review by cacieprins — CHANGES_REQUESTED** [review]
The dependency swap is good, but the `{ subdomain, domain, tld }` compatibility shim is the wrong abstraction. 

tldts already exposes subdomain, domain (eTLD+1), domainWithoutSuffix, and publicSuffix. It also includes IP detection, etc.

We should normalize input (trim, leading dots, allowPrivateDomains) and use those fields directly from cors/cookie logic.

This replaces logic that maps into the old fork shape and re-joins domain + tld, and it will delete most of parse-domain.ts. As a bonus, it makes the super-domain / same-site logic obvious: super-domain is `parsed.domain`, not `compact([domain, tld]).join('.')`.

**Comment by jennifer-shehane** [comment]
@cacieprins This was meant to be a stepping stone instead of a rewrite. It's intention is to purely remove the outside dependency and not change logic / behavior / how we do anything. I can see that as a followup if we can just remove the outside dependency piece so the other repo can be shelved forever.

**Review by cacieprins — APPROVED** [review]

**Comment by jennifer-shehane** [comment]
Archived https://github.com/cypress-io/parse-domain repo
